nupm/install.nu

```diff
@@ -23,107 +23,161 @@ def throw-error [
+        let tgt_path = $scripts_dir | path join $script
 
-    ls --all $source
-    | where {|it| not ($it.type == dir and ($it.name | path parse | get stem) == ".git")}
-    | each {|it|
-        log debug ($it.name | str replace $source "" | str trim --left --char (char path_sep))
-        cp --recursive $it.name $destination
-    }
-}
+        if ($src_path | path type) != file {
+            throw-error $"Script ($src_path) does not exist"
+        }
 
-def install-scripts [path: path, package: record<scripts: list<path>>]: nothing -> nothing {
-    let nupm_bin = $env.NUPM_HOME | path join "bin"
-    mkdir $nupm_bin
-
-    for script in $package.scripts {
-        let script_path = $path | path join $script
-        let name = $script | path basename
-        let destination = $nupm_bin | path join $name
-
-        if ($script_path | path exists) {
-            log debug $"installing script `($name)` to `($destination)`"
-            cp $script_path $destination
-            chmod +x $destination
-        } else {
-            log warning $"($script_path) could not be found, skipping"
+        if ($tgt_path | path type) == file and (not $force) {
+            throw-error ($"Script ($src_path) is already installed as"
+                + $" ($tgt_path)")
         }
+
+        log debug $"installing script `($src_path)` to `($scripts_dir)`"
+        cp $src_path $scripts_dir
     }
 }
 
-# Install a nupm package
-export def main [
-    --path: path  # the path to the local source of the package (defaults to the current directory)
+# Install package from a directory containing 'project.nuon'
+def install-path [
+    pkg_dir: path  # Directory (hopefully) containing 'package.nuon'
+    --force(-f)    # Overwrite already installed package
 ] {
-    if $path == null {
-        throw-error "`nupm install` requires a `--path`"
-    }
+    let pkg_dir = $pkg_dir | path expand --strict
 
-    let package = open-package-file $path
+    let package = open-package-file $pkg_dir
 
     log info $"installing package ($package.name)"
 
     match $package.type {
         "module" => {
-            let destination = $env.NUPM_HOME | path join $package.name
+            let mod_dir = $pkg_dir | path join $package.name
+
+            if ($mod_dir | path type) != dir {
+                throw-error ($"Module package '($package.name)' does not"
+                    + $" contain directory '($package.name)'")
+            }
+
+            let module_dir = module-dir --ensure
+            let destination = $module_dir | path join $package.name
+
+            if $force {
+                rm --recursive --force $destination
+            }
+
+            if ($destination | path type) == dir {
+                throw-error $"Package ($package.name) is already installed"
+            }
 
-            prepare-directory $destination
-            $path | copy-directory-to $destination
+            cp --recursive $mod_dir $module_dir
```

Comment:
is this not copying the whole module to the destination as before? :thinking:

Reply:
It's copying the subdirectory containing the module, not the top directory.

Reply:
ooh yeah, smart, it used to `ls` everything and copy the whole set of files, here `$mod_dir` is the subdirectory which should contain all the module files :+1: